cpp/src/Ice/ConnectionI.cpp

```diff
@@ -2807,6 +2765,41 @@ Ice::ConnectionI::sendMessage(OutgoingMessage& message)
     assert(_state >= StateActive);
     assert(_state < StateClosed);
 
+    bool isHeartbeat = static_cast<uint8_t>(message.stream->b[8]) == IceInternal::validateConnectionMsg;
+    if (!isHeartbeat)
+    {
+        cancelInactivityTimerTask();
```

Comment:
cancelInactivityTimerTask is always no-op for a Reply message because _inactivityTimerTaskScheduled is always false. 

When we send back a Reply (with the connection locked), the dispatch count is just decreased to 0 and there is no way some other heartbeat sending can lock the mutex, see this new _dispatchCount == 0 and schedule the inactivity timer task.